validphys2/src/validphys/convolution.py

```diff
@@ -248,7 +250,23 @@ def linear_predictions(dataset, pdf):
     return _predictions(dataset, pdf, linear_fk_predictions)
 
 
-def fk_predictions(loaded_fk, pdf):
+def determine_pdf(conv_types, pdf, unpolarized_bc):
+    """Look at convolution types, and determine the correct pdf to use
+    for the convolution. The following in principle could accept any
+    number of convolutions.
+    """
+    if unpolarized_bc is not None:
+        map_pdf_to_conv_types = {
+            "PolPDF": pdf,
+            "UnpolPDF": unpolarized_bc,
+        }  # NOTE: add additional PDF type here
+        pdf_list = [map_pdf_to_conv_types.get(i, pdf) for i in conv_types]
+    else:
+        pdf_list = [pdf for _ in conv_types]
+    return pdf_list
```

Comment:
Regarding this, I have two comments:
- actually, we always have to load the entire `unpolarized_bc` replicas set because we need to shift the cv by one standard deviation when imposing the positivity BC  (forgot this myself). And therefore we cannot just load the set as a `t0` PDF.
https://github.com/NNPDF/nnpdf/blob/840ac3e2775929353c7e44be20fd494ff7fb1c10/n3fit/src/n3fit/layers/observable.py#L47-L51

- Isn't the old way actually more modular? In that it is not bounded by `pdf_1` and `pdf_2`. But I agree that this function could be greatly improved. For instance, I don't really like coding manually the mapping of the pdfs from the convolution types.

Reply:
> actually, we always have to load the entire unpolarized_bc replicas set because we need to shift the cv by one standard deviation when imposing the positivity BC (forgot this myself). And therefore we cannot just load the set as a t0 PDF.

That's fine. The `determine_pdf` function can also manipulate the PDF if needed. I would concentrate these manipulations in this function.

The "check that one is central" can also be done at the end "check that one of the two is manipulated". The point is that (the way it is done right now) all these functions work _only_ if the number of members in the unpolarized PDF and the polarized one is the same.

> Isn't the old way actually more modular? In that it is not bounded by pdf_1 and pdf_2. But I agree that this function could be greatly improved. For instance, I don't really like coding manually the mapping of the pdfs from the convolution types.

But it is basically the same. I'm just passing the fktable instead of one of the variables of the fktable, with the idea that eventually there will be also fragmentation functions and that information will be included in the fktable.